### ISSUE TITLE
Trouble connecting to Cassandra cluster

### ISSUE THREAD

**Issue by exnerd** [mesosphere/cassandra-mesos#43]
Hi,

I build the latest cassandra-mesos from the rewrite branch. I started a Cassandra cluster with marathon. I see the Cassandra 'master' coming up and registering the framework 'cassandra.dev-test' in the mesos cluster. Afterwards 3 executor nodes are created. Everything seems fine. The cassandra cluster is up. But when I try to connect with csqlh to the cluster nothing happens. 
What does seems weird is that netstat -anp doesn't show any of the standard cassandra ports. I did add these ports in the mesos-agent config. 
Is my marathon json wrong ?
```
{
  "id": "/cassandra-mesos",
  "instances": 1,
  "cpus": 0.5,
  "mem": 512,
  "ports": [0],
  "uris": [
    "http://some-host/cassandra-0.1.0-SNAPSHOT-jar-with-dependencies.jar",
    "http://some-host/cassandra-executor.jar",
    "http://some-host/cassandra-framework.jar",
    "http://some-host/cassandra.tar.gz",
    "http://some-host/jre-7u76-linux-x64.tar.gz",
    "https://downloads.mesosphere.io/cassandra-mesos/jdk/jdk-7u75-linux-x64.tar.gz"
  ],
  "env": {
    "MESOS_ZK": "zk://zk1:2181,zk2:2181,zk3:2181/mesos",
    "JAVA_OPTS": "-Xms256m -Xmx256m ",
    "CASSANDRA_CLUSTER_NAME": "dev-test",
    "CASSANDRA_ZK": "zk://zk1:2181,zk2:2181,zk3:2181/cassandra-mesos",
    "CASSANDRA_NODE_COUNT": "3",
    "CASSANDRA_RESOURCE_CPU_CORES": "2.0",
    "CASSANDRA_RESOURCE_MEM_MB": "2048",
    "CASSANDRA_RESOURCE_DISK_MB": "2048",
    "CASSANDRA_HEALTH_CHECK_INTERVAL_SECONDS": "60"
  },
  "cmd": "$(pwd)/jdk*/bin/java $JAVA_OPTS -classpath cassandra-framework.jar    io.mesosphere.mesos.frameworks.cassandra.Main"
}
```
Do I need to specify the standard cassandra ports (9042,9160,...) in the marathon json ?
Or is the framework not ready yet to connect to?

Regards,

Benoît

**Comment by BenWhitehead** [comment]
The marathon.json is used to launch the framework which then connects to Mesos using the value specified in the `MESOS_ZK` env variable. When the framework starts it will receive a random port from marathon that it can use to bind it's HTTP server to.

The standard cassandra ports are taken into consideration when the cassandra framework evaluates a resource offer from mesos. If the offer doesn't have all the cpu, memory, disk and ports a cassandra server task takes to run the offer will be declined. In the case that a resource offer has insufficient resources you should see a log statement similar to the following:
```
2015-03-06 11:36:28,580 INFO  [Thread-6] i.m.m.f.cassandra.CassandraCluster   - {offerId:20150306-113343-16777343-5050-10766-O19,hostname:localhost} Insufficient resources in offer: 20150306-113343-16777343-5050-10766-O19. Details: ['Unavailable port 9160(rpc_port). 1005 other ports available.']
```

Logs for the framework can be found in the Mesos sandbox the framework is started in by Mesos in the file `cassandra-mesos.log`.

Due to this constraint in how the Cassandra Framework evaluates resource offers the mesos slave(s) need to be configured to have the standard cassandra ports be part of their resource set. To update the port resources of your slave run the following commands if you're using Mesosphere packages of mesos (.dep or .rpm from our repos)

###### Debian or Ubuntu
```
sudo service mesos-slave stop
echo "ports:[31000-32000,7000-7001,7199-7199,9042-9042,9160-9160]" | sudo tee /etc/mesos-slave/resources
sudo rm -rf /tmp/mesos
sudo service mesos-slave start
```

###### CentOS / Redhat
```
sudo systemctl stop mesos-slave
echo "ports:[31000-32000,7000-7001,7199-7199,9042-9042,9160-9160]" | sudo tee /etc/mesos-slave/resources
sudo rm -rf /tmp/mesos
sudo systemctl start mesos-slave
```

If you're not using the Mesosphere packages, add the following argument to the mesos-slave arguments
```
--resources='ports:[31000-32000,7000-7001,7199-7199,9042-9042,9160-9160]'
```

I hope this helps you get further. If you still run into issues, can you create a gist with the contents of `cassandra-mesos.log` perhaps I can get more info to help diagnose further.

**Comment by exnerd** [comment]
Hi Ben,

I restarted the whole procedure. I removed the framework in mesos. Removed in zookeeper  the cassandra-mesos subtree and restarted the framework.
As asked I made a gist of the first 250 lines of the cassanda-mesos.log https://gist.github.com/exnerd/1403f96953df7b2c164d . 3 executors were started as you can see in the log. Shouldn't the framework stop seeking resources after that? Because as you can see after that the executors are started, the message 'insufficient resources' appears in the cassandra-mesos.log
But what I still don't understand is the following :
The output of the netstat -anp | grep java on the executors gives following result
```
root@mesos-02:~#netstat -anp | grep java
tcp        0      0 0.0.0.0:45858           0.0.0.0:*               LISTEN      30065/java      
tcp        0      0 193.53.25.215:49038     193.53.25.215:5051      ESTABLISHED 30065/java      
```
And the mesos-slave port resources are 
```
[31000-32000,7000-7001,7199-7199,9042-9042,9160-9160]
```
I also added the executor.log of one the 3 executors. 
I'm still figuring this out.

Any help is welcome.

Regards,

Benoît

**Comment by exnerd** [comment]
Something struck me, why would a mesos-slave announce that it hasn't the (port) resources, while nothing was running that used that port. 
In the Mesos UI the mesos-slave(s) showed an active framework with no running task. Those were from a previous runs. I deleted those with
```
  rm -r /tmp/mesos/meta/slaves/latest/framework/<framework id>
```
and restarted the mesos-slaves.

Afterwards a restarted the procedure and now the framework is running 4 task instead of 3 the previous time. I have 3 executors and 1 executor server

**Comment by BenWhitehead** [comment]
Yes you are correct.  The only way that a port wouldn't be part of a resource offer is if it is claimed by another task.  We are currently working on getting lots of improvements into the scheduler so that it can recover more appropriately after restarts and task failures.

**Comment by BenWhitehead** [comment]
@exnerd Where you able to get unblocked, or are you still having issues?

**Comment by exnerd** [comment]
Well Ben,

a bit of both ;-). 
 - Still having in the logs 'Unavailable port 7199' in the cassandra-mesos.log. I checked with netstat -anp, lsof -i on the node, but nothing is listing on that port. I don't know how mesos handles port resources. I haven't found any log or rest-call that shows the port resources per slave or for the entire cluster for that matter. The problem is that you have to assign standard Cassandra ports (9042, 9160,7000,7001) to the mesos slave and since it's cluster other task on that cluster can use those resources. There should be some mechanism that reserves some resources (ports resources in this case) to specific tasks.
 - I had in the framework 4 tasks : 3 executors (cassandra.dev-test.node.[0-2].executor) and 1 server (cassandra.dev-test.node.2.executor.server). It is not clear to me how many tasks there has to be and what there purpose are. 

For now I am waiting for the improvements that you were writing about. I am also evaluating a cassandra cluster with docker images on the mesos cluster.

**Comment by BenWhitehead** [comment]
@exnerd If the port still isn't available the best way to check that the slave has the correct configuration is to check it's state endpoint. Go to `http://slave1.mesos:5051/state.json` to get a json document listing all state of the slave including configuration and resources. You should see something like the json below. Notice the value of `.flags.resources` represents the resources argument passed to the slave when it's process was started. `.resources.ports` should include all the cassandra ports in addition to any other ports you want mesos to be able to provide as resources.

In terms of expected tasks and executors it should look something like this (for a 3 node cluster):
* `cassandra.dev-test.node.0.executor` (actual executor id)
  * `cassandra.dev-test.node.0.executor` (task id) Responsible for collecting slave metadata and tracking executor from UI
  * `cassandra.dev-test.node.0.executor.server`
     Actual cassandra server task that will result is bound ports.
* `cassandra.dev-test.node.1.executor` (actual executor id)
  * `cassandra.dev-test.node.1.executor` (task id)
  * `cassandra.dev-test.node.1.executor.server`
* `cassandra.dev-test.node.2.executor` (actual executor id)
  * `cassandra.dev-test.node.2.executor` (task id)
  * `cassandra.dev-test.node.2.executor.server`

With regard to reserving resources there is an open PR to add mesos role support to the cassandra framework so that an operator could reserve resources (cpus, mem, disk, ports) for a specific role and then start the cassandra framework with that role.

There is active development in mesos to allow frameworks to make this sort of reservation it self which the cassandra framework will utilize in the future so that an operator doesn't have to make the reservation. Once this functionality is in place in mesos we will more aggressively evaluate resource offers and not even start an executor if we're unable to make a resource reservation.

I hope this info helps, and I appreciate your feedback on the framework. I'm hoping that in the next few weeks the code will be in a state that it can replace the `master` branch to start getting more feedback about the framework.

```json
{
    "attributes": {},
    "build_date": "2015-02-27 10:54:31",
    "build_time": 1425063271,
    "build_user": "ben.whitehead",
    "completed_frameworks": [],
    "failed_tasks": 0,
    "finished_tasks": 0,
    "flags": {
        "authenticatee": "crammd5",
        "cgroups_enable_cfs": "false",
        "cgroups_hierarchy": "/sys/fs/cgroup",
        "cgroups_limit_swap": "false",
        "cgroups_root": "mesos",
        "container_disk_watch_interval": "15secs",
        "containerizers": "docker,mesos",
        "default_role": "*",
        "disk_watch_interval": "1mins",
        "docker": "docker",
        "docker_remove_delay": "6hrs",
        "docker_sandbox_directory": "/mnt/mesos/sandbox",
        "docker_stop_timeout": "0ns",
        "enforce_container_disk_quota": "false",
        "executor_registration_timeout": "5mins",
        "executor_shutdown_grace_period": "5secs",
        "frameworks_home": "",
        "gc_delay": "1weeks",
        "gc_disk_headroom": "0.1",
        "hadoop_home": "",
        "help": "false",
        "initialize_driver_logging": "true",
        "isolation": "cgroups/cpu,cgroups/mem",
        "launcher_dir": "/opt/mesos/latest/build/src",
        "log_dir": "/opt/mesos/work/slave/log_dir",
        "logbufsecs": "0",
        "logging_level": "INFO",
        "master": "zk://localhost:2181/mesos",
        "perf_duration": "10secs",
        "perf_interval": "1mins",
        "port": "5051",
        "quiet": "false",
        "recover": "reconnect",
        "recovery_timeout": "15mins",
        "registration_backoff_factor": "1secs",
        "resource_monitoring_interval": "1secs",
        "resources": "ports:[31000-32000,7000-7001,7199-7199,9042-9042,9160-9160]",
        "strict": "true",
        "switch_user": "true",
        "version": "false",
        "work_dir": "/opt/mesos/work/slave/work_dir"
    },
    "frameworks": [],
    "git_branch": "refs/heads/master",
    "git_sha": "2985ae05634038b70f974bbfed6b52fe47231418",
    "hostname": "localhost",
    "id": "20150311-115808-16777343-5050-18022-S0",
    "killed_tasks": 0,
    "log_dir": "/opt/mesos/work/slave/log_dir",
    "lost_tasks": 0,
    "master_hostname": "localhost",
    "pid": "slave(1)@127.0.0.1:5051",
    "resources": {
        "cpus": 8,
        "disk": 224794,
        "mem": 14749,
        "ports": "[31000-32000, 7000-7001, 7199-7199, 9042-9042, 9160-9160]"
    },
    "staged_tasks": 0,
    "start_time": 1426100291.32705,
    "started_tasks": 0,
    "version": "0.23.0"
}
```

**Comment by exnerd** [comment]
Thanks for the explanation, especially the expected tasks and executors can help in debugging the whole setup. I will follow the progress of this project in the coming weeks.
Let me know if I can help with something.

Regards,
exnerd

**Comment by BenWhitehead** [comment]
Sure, thanks for being eager to try it out and providing feedback.
